File: Java/Trapping-Rain-Water.java
public class Solution {
    /**
     * @param heights: an array of integers
     * @return: a integer
     */
    public int trapRainWater(int[] heights) {
        // write your code here
        int maxh = -1, index = 0;
        for (int i = 0; i < heights.length; i ++) {
            if (heights[i] > maxh) {
                index = i;
                maxh = heights[i];
            }
        }
        int res = 0, lefth = -1, righth = -1;
        for (int i = 0; i < index; i ++) {
            if (heights[i] >= lefth) {
                lefth = heights[i];
            } else {
                res += (lefth - heights[i]);
            }
        }
        for (int i = heights.length - 1; i > index; i --) {
            if (heights[i] >= righth) {
                righth = heights[i];
            } else {
                res += (righth - heights[i]);
            }
        }
        return res;
    }
}

File: Python/Trapping-Rain-Water.py
class Solution:
    # @param heights: a list of integers
    # @return: a integer
    def trapRainWater(self, heights):
        # write your code here
        if len(heights) == 0:
            return 0
        maxh = heights[0]
        cur = 0
        for i in range(len(heights)):
            if heights[i] > maxh:
                maxh = heights[i]
                cur = i
        res = 0
        pre = heights[0]
        for i in range(cur):
            res += max(0, pre - heights[i])
            pre = max(pre, heights[i])
        pre = heights[-1]
        index = len(heights) - 1
        while index > cur:
            res += max(0, pre - heights[index])
            pre = max(pre, heights[index])
            index -= 1
        return res
